File: tools/s07/memory-profile/go/main.go
// Diagnostic memory adapter, built inside an exact pinned Go source export.
// Profiles/census/checkpoint pauses deliberately perturb execution and cannot
// replace the ordinary allocation, timing, or lifetime-RSS acceptance samples.
package main

import (
	"bufio"
	"crypto/sha256"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"runtime/debug"
	"runtime/pprof"
	"sort"
	"strconv"
	"sync"
	"time"

	"github.com/microsoft/TypeScript/tsc/internal/ast"
	"github.com/microsoft/TypeScript/tsc/internal/binder"
	"github.com/microsoft/TypeScript/tsc/internal/core"
	"github.com/microsoft/TypeScript/tsc/internal/parser"
	"github.com/microsoft/TypeScript/tsc/internal/tspath"
	"github.com/microsoft/TypeScript/tsc/internal/vfs/osvfs"
)

// The pinned default is 512 KiB. This diagnostic uses 64 KiB to improve
// attribution of smaller sites while retaining statistical sampling; all
// pipeline work happens after this single early rate assignment.
const memoryProfileRate = 64 * 1024

func init() { runtime.MemProfileRate = memoryProfileRate }

type Input struct {
	Filename   string `json:"filename"`
	Path       string `json:"path"`
	Local      string `json:"local"`
	ScriptKind int32  `json:"script_kind"`
	JSX        bool   `json:"jsx"`
	Force      bool   `json:"force"`
}

type Loaded struct {
	text    string
	options ast.SourceFileParseOptions
	kind    core.ScriptKind
}

type Snapshot struct {
	Name       string           `json:"name"`
	UnixTimeNs int64            `json:"unix_time_ns"`
	Stats      runtime.MemStats `json:"mem_stats"`
}

type Profile struct {
	Name         string   `json:"name"`
	HeapPath     string   `json:"heap_path"`
	HeapSHA256   string   `json:"heap_sha256"`
	AllocsPath   string   `json:"allocs_path"`
	AllocsSHA256 string   `json:"allocs_sha256"`
	Before       Snapshot `json:"before"`
	After        Snapshot `json:"after"`
	WallTimeNs   int64    `json:"wall_time_ns"`
}

type Worker struct {
	Worker     int   `json:"worker"`
	Files      int   `json:"files"`
	ParseCalls int   `json:"parse_calls"`
	BindCalls  int   `json:"bind_calls"`
	ParseNs    int64 `json:"parse_ns"`
	BindNs     int64 `json:"bind_ns"`
}

type Census struct {
	Files                        int                  `json:"files"`
	MaximumFilePointerIdentities uint64               `json:"maximum_file_pointer_identities"`
	Categories                   ast.MemoryFileCensus `json:"categories"`
	Before                       Snapshot             `json:"before"`
	After                        Snapshot             `json:"after"`
	WallTimeNs                   int64                `json:"wall_time_ns"`
}

type Report struct {
	Version           int        `json:"version"`
	Operation         string     `json:"operation"`
	DiagnosticOnly    bool       `json:"diagnostic_only"`
	MemProfileRate    int        `json:"mem_profile_rate"`
	LoadedInputSHA256 string     `json:"loaded_input_sha256"`
	Workers           int        `json:"workers"`
	Files             int        `json:"files"`
	LoadedBytes       int        `json:"loaded_bytes"`
	Nodes             int        `json:"nodes"`
	Symbols           int        `json:"symbols"`
	ParseDiagnostics  int        `json:"parse_diagnostics"`
	BindDiagnostics   int        `json:"bind_diagnostics"`
	WallTimeNs        int64      `json:"wall_time_ns"`
	AllocatedBytes    uint64     `json:"allocated_bytes"`
	WorkerTimes       []Worker   `json:"worker_times"`
	GOMAXPROCS        int        `json:"gomaxprocs"`
	GOGC              int        `json:"gogc"`
	CPUCapacity       int        `json:"cpu_capacity"`
	GoroutinesReady   int        `json:"goroutines_ready"`
	PipelineBefore    Snapshot   `json:"pipeline_before"`
	Snapshots         []Snapshot `json:"snapshots"`
	Profiles          []Profile  `json:"profiles"`
	Census            Census     `json:"census"`
	Limitations       []string   `json:"limitations"`
}

func loadedDigest(inputs []Loaded) string {
	digest := sha256.New()
	digest.Write([]byte("S07-loaded-inputs-v1\x00"))
	word := make([]byte, 8)
	binary.BigEndian.PutUint64(word, uint64(len(inputs)))
	digest.Write(word)
	for _, input := range inputs {
		for _, value := range []string{input.options.FileName, string(input.options.Path)} {
			binary.BigEndian.PutUint64(word, uint64(len(value)))
			digest.Write(word)
			digest.Write([]byte(value))
		}
		binary.BigEndian.PutUint32(word[:4], uint32(input.kind))
		digest.Write(word[:4])
		bits := []byte{0, 0}
		if input.options.ExternalModuleIndicatorOptions.JSX {
			bits[0] = 1
		}
		if input.options.ExternalModuleIndicatorOptions.Force {
			bits[1] = 1
		}
		digest.Write(bits)
		binary.BigEndian.PutUint64(word, uint64(len(input.text)))
		digest.Write(word)
		source := sha256.Sum256([]byte(input.text))
		digest.Write(source[:])
	}
	return fmt.Sprintf("%x", digest.Sum(nil))
}

func snapshot(name string) Snapshot {
	result := Snapshot{Name: name, UnixTimeNs: time.Now().UnixNano()}
	runtime.ReadMemStats(&result.Stats)
	return result
}

func checkpoint(value Snapshot, reader *bufio.Reader, writer *bufio.Writer) {
	if err := json.NewEncoder(writer).Encode(struct {
		Checkpoint string   `json:"checkpoint"`
		Snapshot   Snapshot `json:"snapshot"`
	}{value.Name, value}); err != nil {
		panic(err)
	}
	if err := writer.Flush(); err != nil {
		panic(err)
	}
	line, err := reader.ReadString('\n')
	if err != nil {
		panic(fmt.Sprintf("checkpoint %s resume failed: %v", value.Name, err))
	}
	if line != "\n" && line != "\r\n" {
		panic("checkpoint resume must be one empty line")
	}
}

func fileDigest(path string) string {
	file, err := os.Open(path)
	if err != nil {
		panic(err)
	}
	defer file.Close()
	digest := sha256.New()
	if _, err := io.Copy(digest, file); err != nil {
		panic(err)
	}
	return fmt.Sprintf("%x", digest.Sum(nil))
}

func profiles(prefix, name string) Profile {
	result := Profile{Name: name, HeapPath: prefix + "-" + name + ".heap.pprof",
		AllocsPath: prefix + "-" + name + ".allocs.pprof", Before: snapshot(name + "_profile_before")}
	started := time.Now()
	for _, output := range []struct{ kind, path string }{{"heap", result.HeapPath}, {"allocs", result.AllocsPath}} {
		file, err := os.Create(output.path)
		if err != nil {
			panic(err)
		}
		profile := pprof.Lookup(output.kind)
		if profile == nil {
			panic("missing runtime memory profile: " + output.kind)
		}
		if err := profile.WriteTo(file, 0); err != nil {
			file.Close()
			panic(err)
		}
		if err := file.Close(); err != nil {
			panic(err)
		}
	}
	result.HeapSHA256 = fileDigest(result.HeapPath)
	result.AllocsSHA256 = fileDigest(result.AllocsPath)
	result.WallTimeNs = time.Since(started).Nanoseconds()
	result.After = snapshot(name + "_profile_after")
	return result
}

func main() {
	if len(os.Args) != 4 {
		panic("usage: memory-profile INPUTS.json WORKERS (1 or 8) OUTPUT_PREFIX")
	}
	workers, err := strconv.Atoi(os.Args[2])
	if err != nil || workers != 1 && workers != 8 {
		panic("worker count must be 1 or 8")
	}
	if os.Args[3] == "" {
		panic("empty output prefix")
	}
	file, err := os.Open(os.Args[1])
	if err != nil {
		panic(err)
	}
	decoder := json.NewDecoder(file)
	decoder.DisallowUnknownFields()
	var requests []Input
	if err := decoder.Decode(&requests); err != nil {
		panic(err)
	}
	var extra any
	if err := decoder.Decode(&extra); err != io.EOF {
		panic("workload must contain exactly one JSON value")
	}
	if err := file.Close(); err != nil {
		panic(err)
	}
	if len(requests) == 0 {
		panic("empty parse-and-bind workload")
	}
	runtime.GOMAXPROCS(workers)
	debug.SetGCPercent(100)
	inputs := make([]Loaded, len(requests))
	loadedBytes := 0
	for index, request := range requests {
		text, ok := osvfs.FS().ReadFile(request.Local)
		if !ok {
			panic("could not load workload file: " + request.Local)
		}
		inputs[index] = Loaded{text, ast.SourceFileParseOptions{FileName: request.Filename, Path: tspath.Path(request.Path),
			ExternalModuleIndicatorOptions: ast.ExternalModuleIndicatorOptions{JSX: request.JSX, Force: request.Force}}, core.ScriptKind(request.ScriptKind)}
		loadedBytes += len(text)
	}
	identity := loadedDigest(inputs)
	prefix, err := filepath.Abs(os.Args[3])
	if err != nil {
		panic(err)
	}
	if err := os.MkdirAll(filepath.Dir(prefix), 0o755); err != nil {
		panic(err)
	}
	output, err := os.Create(prefix + "-report.json")
	if err != nil {
		panic(err)
	}
	defer output.Close()
	reader, writer := bufio.NewReader(os.Stdin), bufio.NewWriterSize(os.Stdout, 64*1024)
	channels := make([]chan int, workers)
	results := make([][]*ast.SourceFile, workers)
	statistics := make([]Worker, workers)
	failures := make([]any, workers)
	ready, done := sync.WaitGroup{}, sync.WaitGroup{}
	ready.Add(workers)
	done.Add(workers)
	start := make(chan struct{})
	for worker := range workers {
		channels[worker] = make(chan int, workers)
		results[worker] = make([]*ast.SourceFile, 0, (len(inputs)+workers-1)/workers)
		statistics[worker].Worker = worker
		go func(worker int) {
			defer done.Done()
			ready.Done()
			<-start
			for index := range channels[worker] {
				if failures[worker] != nil {
					continue
				}
				func() {
					complete := false
					defer func() {
						failure := recover()
						if !complete {
							failures[worker] = fmt.Sprintf("file %s: %v\n%s", inputs[index].options.FileName, failure, debug.Stack())
						}
					}()
					input := inputs[index]
					started := time.Now()
					file := parser.ParseSourceFile(input.options, input.text, input.kind)
					statistics[worker].ParseNs += time.Since(started).Nanoseconds()
					statistics[worker].ParseCalls++
					started = time.Now()
					binder.BindSourceFile(file)
					statistics[worker].BindNs += time.Since(started).Nanoseconds()
					statistics[worker].BindCalls++
					results[worker] = append(results[worker], file)
					statistics[worker].Files++
					complete = true
				}()
			}
		}(worker)
	}
	ready.Wait()
	report := Report{Version: 1, Operation: "parse_bind_memory_profile", DiagnosticOnly: true,
		MemProfileRate: memoryProfileRate, LoadedInputSHA256: identity, Workers: workers, LoadedBytes: loadedBytes,
		GOMAXPROCS: runtime.GOMAXPROCS(0), GOGC: 100, CPUCapacity: runtime.NumCPU(), GoroutinesReady: runtime.NumGoroutine(),
		WorkerTimes: statistics, Snapshots: make([]Snapshot, 0, 5), Profiles: make([]Profile, 0, 4)}
	report.Profiles = append(report.Profiles, profiles(prefix, "pre_pipeline"))
	before := snapshot("pre_pipeline")
	report.Snapshots = append(report.Snapshots, before)
	checkpoint(before, reader, writer)
	// Reset after all pre-pipeline profile/checkpoint serialization. Exact
	// TotalAlloc delta therefore excludes that diagnostic preparation.
	report.PipelineBefore = snapshot("pipeline_before_barrier")
	started := time.Now()
	close(start)
	for index := range inputs {
		channels[index%workers] <- index
	}
	for _, channel := range channels {
		close(channel)
	}
	done.Wait()
	report.WallTimeNs = time.Since(started).Nanoseconds()
	endpoint := snapshot("retained_endpoint")
	report.Snapshots = append(report.Snapshots, endpoint)
	checkpoint(endpoint, reader, writer)
	for worker, files := range results {
		if failures[worker] != nil {
			panic(failures[worker])
		}
		expected := len(inputs) / workers
		if worker < len(inputs)%workers {
			expected++
		}
		stats := statistics[worker]
		if len(files) != expected || stats.Files != expected || stats.ParseCalls != expected || stats.BindCalls != expected {
			panic(fmt.Sprintf("worker %d workload count mismatch", worker))
		}
		for _, file := range files {
			report.Files++
			report.Nodes += file.NodeCount
			report.Symbols += file.SymbolCount
			report.ParseDiagnostics += len(file.Diagnostics())
			report.BindDiagnostics += len(file.BindDiagnostics())
		}
	}
	if report.Files != len(inputs) || endpoint.Stats.TotalAlloc < report.PipelineBefore.Stats.TotalAlloc {
		panic("workload/accounting mismatch")
	}
	report.AllocatedBytes = endpoint.Stats.TotalAlloc - report.PipelineBefore.Stats.TotalAlloc
	report.Profiles = append(report.Profiles, profiles(prefix, "retained_endpoint"))
	// The pinned runtime documents memory profiles as up to two GC cycles old.
	// These deliberate cycles are separate from the ordinary retained endpoint.
	runtime.GC()
	runtime.GC()
	retainedGC := snapshot("retained_after_gc")
	report.Snapshots = append(report.Snapshots, retainedGC)
	checkpoint(retainedGC, reader, writer)
	report.Profiles = append(report.Profiles, profiles(prefix, "retained_after_gc"))
	report.Census = census(results)
	runtime.KeepAlive(inputs)
	runtime.KeepAlive(results)
	inputs, results, requests, channels, failures = nil, nil, nil, nil, nil
	retired := snapshot("post_retirement")
	report.Snapshots = append(report.Snapshots, retired)
	checkpoint(retired, reader, writer)
	runtime.GC()
	runtime.GC()
	retiredGC := snapshot("post_retirement_after_gc")
	report.Snapshots = append(report.Snapshots, retiredGC)
	checkpoint(retiredGC, reader, writer)
	report.Profiles = append(report.Profiles, profiles(prefix, "post_retirement_after_gc"))
	report.Limitations = []string{
		"Diagnostic only: 64KiB sampled heap/alloc profiles, checkpoint pauses and explicit GCs perturb execution; not acceptance allocation/RSS/time measurements.",
		"Exact pipeline allocated_bytes is MemStats.TotalAlloc delta after pre-pipeline profile/checkpoint overhead and before endpoint serialization. Sampled alloc_space estimates use allocation stacks and may differ; no forced reconciliation.",
		"Heap and allocs profiles contain the same four sample types with different defaults. Ordinary endpoint profiles can lag up to two GC cycles. The separate retained_after_gc profiles follow two requested GC cycles while roots remain live.",
		"Checkpoint MemStats precedes its own JSON serialization and pause; parent OS snapshots occur afterward. Snapshot/profile/checkpoint bookkeeping remains visible in whole-process heap data.",
		"Census runs after retained heap snapshots. Its measured allocations and temporary pointer maps must not be attributed to compiler work or interpreted as compiler peak RSS.",
		"Census counts reachable logical objects, including metadata and synthetic flow nodes, not all historically allocated factory nodes/symbols or every element kept alive by an arena backing array.",
		"Concrete Go AST object sizes include the embedded Node header and binding fields. node_header_contained_bytes is a contained subcategory and must not be added to concrete object bytes.",
		"Slice capacity ranges are unioned within each file and element type. Arena backing slack outside visible slices, map buckets/control bytes/spare capacity, size-class rounding and allocator/GC metadata remain unmeasured by the census.",
		"String counts include descriptors' selected bytes, not necessarily separate allocations. Source-backed aliases and per-file visible range unions are reported; other ranges include read-only/global strings and are not a heap-byte subtotal. Shared globals can recur across files.",
		"Post-retirement snapshots release compiler roots and preload buffers; they still include runtime/profiler/report metadata and allocator/OS footprint caused by prior census scratch allocations. The first snapshot requests no GC; the second follows two explicit GCs, which do not require immediate OS page release. Automatic collections may also occur.",
	}
	if err := json.NewEncoder(output).Encode(report); err != nil {
		panic(err)
	}
	if err := output.Close(); err != nil {
		panic(err)
	}
	if err := json.NewEncoder(writer).Encode(map[string]any{"complete": true, "diagnostic_only": true,
		"report": prefix + "-report.json", "report_sha256": fileDigest(prefix + "-report.json"),
		"files": report.Files, "loaded_input_sha256": identity}); err != nil {
		panic(err)
	}
	if err := writer.Flush(); err != nil {
		panic(err)
	}
}

func census(files [][]*ast.SourceFile) Census {
	result := Census{Before: snapshot("census_before")}
	started := time.Now()
	objects := make(map[string]ast.MemoryTypeCensus)
	slices := make(map[string]ast.MemorySliceCensus)
	maps := make(map[string]ast.MemoryMapCensus)
	result.Categories.NodeKinds = make(map[string]uint64)
	result.Categories.SkippedExternalPointerFields = make(map[string]uint64)
	for _, group := range files {
		for _, file := range group {
			current := ast.MemoryProfileCensus(file)
			result.Files++
			result.MaximumFilePointerIdentities = max(result.MaximumFilePointerIdentities, current.PointerIdentities)
			result.Categories.PointerIdentities += current.PointerIdentities
			result.Categories.SourceTextBytes += current.SourceTextBytes
			result.Categories.NodeHeaderContainedBytes += current.NodeHeaderContainedBytes
			for name, count := range current.NodeKinds {
				result.Categories.NodeKinds[name] += count
			}
			for name, count := range current.SkippedExternalPointerFields {
				result.Categories.SkippedExternalPointerFields[name] += count
			}
			for _, item := range current.Objects {
				old, exists := objects[item.Type]
				if exists && (old.Size != item.Size || old.Category != item.Category) {
					panic("census type layout changed")
				}
				item.Count += old.Count
				item.LogicalBytes += old.LogicalBytes
				objects[item.Type] = item
			}
			for _, item := range current.Slices {
				old := slices[item.Type]
				item.Headers += old.Headers
				item.NilHeaders += old.NilHeaders
				item.LengthElements += old.LengthElements
				item.CapacityElements += old.CapacityElements
				item.VisibleCapacityUnion += old.VisibleCapacityUnion
				slices[item.Type] = item
			}
			for _, item := range current.Maps {
				old := maps[item.Type]
				item.Maps += old.Maps
				item.Entries += old.Entries
				item.EntryLogicalBytes += old.EntryLogicalBytes
				maps[item.Type] = item
			}
			strings := &result.Categories.Strings
			strings.Fields += current.Strings.Fields
			strings.EmptyFields += current.Strings.EmptyFields
			strings.LogicalBytes += current.Strings.LogicalBytes
			strings.SourceBackedFields += current.Strings.SourceBackedFields
			strings.SourceBackedLogicalBytes += current.Strings.SourceBackedLogicalBytes
			strings.SourceVisibleUnionBytes += current.Strings.SourceVisibleUnionBytes
			strings.OtherVisibleUnionBytes += current.Strings.OtherVisibleUnionBytes
		}
	}
	for _, item := range objects {
		result.Categories.Objects = append(result.Categories.Objects, item)
	}
	for _, item := range slices {
		result.Categories.Slices = append(result.Categories.Slices, item)
	}
	for _, item := range maps {
		result.Categories.Maps = append(result.Categories.Maps, item)
	}
	sort.Slice(result.Categories.Objects, func(i, j int) bool { return result.Categories.Objects[i].Type < result.Categories.Objects[j].Type })
	sort.Slice(result.Categories.Slices, func(i, j int) bool { return result.Categories.Slices[i].Type < result.Categories.Slices[j].Type })
	sort.Slice(result.Categories.Maps, func(i, j int) bool { return result.Categories.Maps[i].Type < result.Categories.Maps[j].Type })
	result.WallTimeNs = time.Since(started).Nanoseconds()
	result.After = snapshot("census_after")
	return result
}

File: tools/s07/memory-profile/rust/src/main.rs
//! Staged-only retained memory and allocation diagnostics for frozen S07 work.
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::{
    fs,
    hint::black_box,
    io::{self, Write},
    path::Path,
    sync::{mpsc, Barrier},
    thread,
    time::Instant,
};
use ts_ast::{
    AstFile, BoundFile, ExternalModuleIndicatorOptions, JsString, NodeId, ParsedFile,
    SourceFileParseOptions,
};
use ts_jsstring::SourceText;

#[cfg(not(feature = "sites"))]
#[global_allocator]
static ALLOCATOR: cap::Cap<mimalloc::MiMalloc> = cap::Cap::new(mimalloc::MiMalloc, usize::MAX);
#[cfg(feature = "sites")]
#[global_allocator]
static ALLOCATOR: cap::Cap<alloc_tracker::Allocator<mimalloc::MiMalloc>> = cap::Cap::new(
    alloc_tracker::Allocator::new(mimalloc::MiMalloc),
    usize::MAX,
);
#[cfg(feature = "sites")]
use ts_jsstring::memory_sites::{self, Phase};

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Input {
    filename: String,
    path: String,
    local: String,
    script_kind: i32,
    jsx: bool,
    force: bool,
}
struct Loaded {
    source: SourceText,
    options: SourceFileParseOptions,
    script_kind: ts_core::ScriptKind,
}
#[derive(Default, Serialize)]
struct WorkerTimes {
    parse_ns: u128,
    publish_ns: u128,
    bind_ns: u128,
    parse_allocated_bytes: usize,
    publish_allocated_bytes: usize,
    bind_allocated_bytes: usize,
    parse_live_growth_bytes: isize,
    publish_live_growth_bytes: isize,
    bind_live_growth_bytes: isize,
}
#[derive(Serialize)]
struct Report {
    version: u32,
    runtime: &'static str,
    diagnostic_only: bool,
    workers: usize,
    files: usize,
    loaded_bytes: usize,
    loaded_input_sha256: String,
    nodes: i64,
    symbols: isize,
    parse_diagnostics: usize,
    bind_diagnostics: usize,
    pipeline_wall_ns: u128,
    worker_times: Vec<WorkerTimes>,
    elapsed_worker_totals: WorkerTimes,
    timer_domain: &'static str,
    pre_pipeline: Memory,
    retained_endpoint: Memory,
    pipeline_allocated_bytes: usize,
    pipeline_live_growth_bytes: isize,
    pipeline_superseded_or_freed_requested_bytes: isize,
    allocation_domain: &'static str,
}

#[derive(Clone, Copy, Serialize)]
struct Memory {
    live_requested_bytes: usize,
    total_requested_bytes: usize,
    peak_requested_bytes: usize,
}
fn memory() -> Memory {
    Memory {
        live_requested_bytes: ALLOCATOR.allocated(),
        total_requested_bytes: ALLOCATOR.total_allocated(),
        peak_requested_bytes: ALLOCATOR.max_allocated(),
    }
}
fn checkpoint(name: &str, snapshot: Memory) -> io::Result<()> {
    println!(
        "{}",
        serde_json::json!({"checkpoint": name, "memory": snapshot})
    );
    io::stdout().flush()?;
    let mut input = String::new();
    if io::stdin().read_line(&mut input)? == 0 || input != "\n" {
        return Err(io::Error::new(
            io::ErrorKind::UnexpectedEof,
            "checkpoint requires one newline",
        ));
    }
    Ok(())
}

#[inline(never)]
fn profile_preload(path: &Path) -> Result<Vec<Loaded>, Box<dyn std::error::Error>> {
    #[cfg(feature = "sites")]
    let _phase = memory_sites::phase(Phase::Preload);
    let inputs: Vec<Input> = serde_json::from_slice(&fs::read(path)?)?;
    if inputs.is_empty() {
        return Err("empty workload".into());
    }
    inputs
        .into_iter()
        .map(|input| {
            Ok(Loaded {
                source: SourceText::from_bytes(fs::read(input.local)?),
                options: SourceFileParseOptions {
                    file_name: JsString::from_bytes(input.filename.into_bytes()),
                    path: JsString::from_bytes(input.path.into_bytes()),
                    external_module_indicator_options: ExternalModuleIndicatorOptions {
                        jsx: input.jsx,
                        force: input.force,
                    },
                },
                script_kind: ts_core::ScriptKind(input.script_kind),
            })
        })
        .collect()
}

fn loaded_digest(inputs: &[Loaded]) -> String {
    let mut digest = Sha256::new();
    digest.update(b"S07-loaded-inputs-v1\0");
    digest.update((inputs.len() as u64).to_be_bytes());
    for input in inputs {
        for value in [
            input.options.file_name.as_bytes(),
            input.options.path.as_bytes(),
        ] {
            digest.update((value.len() as u64).to_be_bytes());
            digest.update(value);
        }
        digest.update(input.script_kind.0.to_be_bytes());
        digest.update([
            u8::from(input.options.external_module_indicator_options.jsx),
            u8::from(input.options.external_module_indicator_options.force),
        ]);
        digest.update((input.source.len() as u64).to_be_bytes());
        digest.update(Sha256::digest(input.source.as_bytes()));
    }
    format!("{:x}", digest.finalize())
}

// Named, non-inlined frames preserve the original interleaved operation order.
// black_box prevents a tail call from erasing the phase frame in sampled stacks.
#[inline(never)]
fn profile_parse(input: &Loaded) -> ParsedFile {
    #[cfg(feature = "sites")]
    let _phase = memory_sites::phase(Phase::Parse);
    black_box(ts_parser::parse_source_file(
        input.source.clone(),
        input.script_kind,
        input.options.clone(),
    ))
}

#[inline(never)]
fn profile_publish(parsed: ParsedFile) -> (AstFile, NodeId) {
    #[cfg(feature = "sites")]
    let _phase = memory_sites::phase(Phase::Publish);
    let source = parsed.root();
    black_box((parsed.publish_unbound(), source))
}

#[inline(never)]
fn profile_bind(file: AstFile, source: NodeId) -> BoundFile {
    #[cfg(feature = "sites")]
    let _phase = memory_sites::phase(Phase::Bind);
    let bound = ts_binder::bind_source_file(&file, source).expect("workload binding must complete");
    drop(file);
    black_box(bound)
}

#[inline(never)]
fn measure_pipeline(
    senders: Vec<mpsc::SyncSender<usize>>,
    count: usize,
    finished: &Barrier,
) -> u128 {
    let start = Instant::now();
    for index in 0..count {
        senders[index % senders.len()]
            .send(index)
            .expect("worker drains every input");
    }
    drop(senders);
    finished.wait();
    black_box(start.elapsed().as_nanos())
}

#[inline(never)]
fn profile_retirement(files: Vec<Vec<BoundFile>>) {
    #[cfg(feature = "sites")]
    let _phase = memory_sites::phase(Phase::Retire);
    drop(black_box(files));
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<_> = std::env::args_os().collect();
    if args.len() != 4 {
        return Err("usage: ts_memory_profile INPUTS.json WORKERS OUTPUT_PREFIX".into());
    }
    let workers: usize = args[2].to_str().ok_or("invalid workers")?.parse()?;
    if !matches!(workers, 1 | 8) || thread::available_parallelism()?.get() < workers {
        return Err("profiling requires 1 or 8 available workers".into());
    }
    #[cfg(feature = "sites")]
    memory_sites::initialize();
    let prefix = Path::new(&args[3]);
    let inputs = profile_preload(Path::new(&args[1]))?;
    let file_count = inputs.len();
    let loaded_bytes = inputs.iter().map(|input| input.source.len()).sum();
    let loaded_input_sha256 = loaded_digest(&inputs);
    let ready = Barrier::new(workers + 1);
    let finished = Barrier::new(workers + 1);
    let release = Barrier::new(workers + 1);
    thread::scope(|scope| -> Result<(), Box<dyn std::error::Error>> {
        let mut senders = Vec::with_capacity(workers);
        let mut handles = Vec::with_capacity(workers);
        for _ in 0..workers {
            let (send, receive) = mpsc::sync_channel::<usize>(workers);
            senders.push(send);
            let inputs = &inputs;
            let (ready, finished, release) = (&ready, &finished, &release);
            handles.push(ts_parser::spawn_parser_worker(scope, move || {
                let mut roots = Vec::with_capacity(file_count.div_ceil(workers));
                let mut times = WorkerTimes::default();
                let mut failure = None;
                #[cfg(feature = "sites")]
                memory_sites::initialize_thread();
                ready.wait();
                while let Ok(index) = receive.recv() {
                    if failure.is_some() {
                        continue;
                    }
                    let outcome = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
                        let live_before = ALLOCATOR.allocated();
                        let allocated_before = ALLOCATOR.total_allocated();
                        let start = Instant::now();
                        let parsed = profile_parse(&inputs[index]);
                        times.parse_ns += start.elapsed().as_nanos();
                        let after_parse = ALLOCATOR.total_allocated();
                        let live_after_parse = ALLOCATOR.allocated();
                        if workers == 1 {
                            times.parse_live_growth_bytes +=
                                live_after_parse as isize - live_before as isize;
                        }
                        if workers == 1 {
                            times.parse_allocated_bytes += after_parse - allocated_before;
                        }
                        let start = Instant::now();
                        let (file, source) = profile_publish(parsed);
                        times.publish_ns += start.elapsed().as_nanos();
                        let after_publish = ALLOCATOR.total_allocated();
                        let live_after_publish = ALLOCATOR.allocated();
                        if workers == 1 {
                            times.publish_live_growth_bytes +=
                                live_after_publish as isize - live_after_parse as isize;
                        }
                        if workers == 1 {
                            times.publish_allocated_bytes += after_publish - after_parse;
                        }
                        let start = Instant::now();
                        let bound = profile_bind(file, source);
                        times.bind_ns += start.elapsed().as_nanos();
                        if workers == 1 {
                            times.bind_allocated_bytes +=
                                ALLOCATOR.total_allocated() - after_publish;
                            times.bind_live_growth_bytes +=
                                ALLOCATOR.allocated() as isize - live_after_publish as isize;
                        }
                        bound
                    }));
                    match outcome {
                        Ok(file) => roots.push(file),
                        Err(panic) => failure = Some(panic),
                    }
                }
                finished.wait();
                release.wait();
                if let Some(panic) = failure {
                    std::panic::resume_unwind(panic);
                }
                (roots, times)
            })?);
        }
        ready.wait();
        checkpoint("pre_pipeline", memory())?;
        let pre_pipeline = memory();
        let pipeline_wall_ns = measure_pipeline(senders, file_count, &finished);
        let retained_endpoint = memory();
        checkpoint("retained_endpoint", retained_endpoint)?;
        release.wait();
        let mut files = Vec::with_capacity(workers);
        let mut worker_times = Vec::with_capacity(workers);
        let mut totals = WorkerTimes::default();
        for handle in handles {
            let (roots, times) = handle
                .join()
                .unwrap_or_else(|panic| std::panic::resume_unwind(panic));
            totals.parse_ns += times.parse_ns;
            totals.publish_ns += times.publish_ns;
            totals.bind_ns += times.bind_ns;
            totals.parse_allocated_bytes += times.parse_allocated_bytes;
            totals.publish_allocated_bytes += times.publish_allocated_bytes;
            totals.bind_allocated_bytes += times.bind_allocated_bytes;
            totals.parse_live_growth_bytes += times.parse_live_growth_bytes;
            totals.publish_live_growth_bytes += times.publish_live_growth_bytes;
            totals.bind_live_growth_bytes += times.bind_live_growth_bytes;
            worker_times.push(times);
            files.push(roots);
        }
        let mut report = Report {
            version: 1,
            runtime: "rust",
            diagnostic_only: true,
            workers,
            files: 0,
            loaded_bytes,
            loaded_input_sha256,
            nodes: 0,
            symbols: 0,
            parse_diagnostics: 0,
            bind_diagnostics: 0,
            pipeline_wall_ns,
            worker_times,
            elapsed_worker_totals: totals,
            pre_pipeline,
            retained_endpoint,
            pipeline_allocated_bytes: retained_endpoint.total_requested_bytes - pre_pipeline.total_requested_bytes,
            pipeline_live_growth_bytes: retained_endpoint.live_requested_bytes as isize - pre_pipeline.live_requested_bytes as isize,
            pipeline_superseded_or_freed_requested_bytes: (retained_endpoint.total_requested_bytes - pre_pipeline.total_requested_bytes) as isize - (retained_endpoint.live_requested_bytes as isize - pre_pipeline.live_requested_bytes as isize),
            allocation_domain: "cap requested-byte counters, native mimalloc alloc/zero/realloc forwarding; successful realloc charges full new requested size, including in-place/shrink; churn subtraction counts freed OR superseded requests, not necessarily physical copies; phase counters are global request deltas measured only for one worker (zeros otherwise), so can include main-thread/channel allocations; separate sites build provides thread-local scopes",
            timer_domain:
                "sum of per-file elapsed worker timers; includes descheduling; not CPU time",
        };
        for file in files.iter().flatten() {
            let view = file.view();
            let source = view.source_file()?;
            report.files += 1;
            report.nodes += source.node_count;
            report.symbols += view.result().symbol_count();
            report.parse_diagnostics += source.diagnostics.len();
            report.bind_diagnostics += view.result().diagnostics().len();
        }
        if report.files != file_count {
            return Err("incomplete workload".into());
        }
        let census_before = memory();
        let census_start = Instant::now();
        let mut census = ts_jsstring::census::Collector::default();
        for file in files.iter().flatten() {
            ts_ast::add_retained_file(file, &mut census);
        }
        use ts_jsstring::census::Walk;
        for input in &inputs {
            input.source.walk(&mut census, "preload.source");
            input
                .options
                .file_name
                .walk(&mut census, "preload.file_name");
            input.options.path.walk(&mut census, "preload.path");
        }
        census.vector(&inputs, "preload.records");
        for roots in &files {
            census.vector(roots, "driver.root_vectors");
        }
        census.vector(&files, "driver.worker_roots");
        let census_after = memory();
        let census_wall_ns = census_start.elapsed().as_nanos();
        fs::write(format!("{}-census.json", prefix.display()), census.json())?;
        drop(census);
        let mut output = serde_json::to_value(&report)?;
        output["census_cost"] = serde_json::json!({"before": census_before, "after": census_after, "wall_ns": census_wall_ns});
        fs::write(
            format!("{}-report.json", prefix.display()),
            serde_json::to_vec_pretty(&output)?,
        )?;
        black_box(&files);
        profile_retirement(files);
        #[cfg(feature = "sites")]
        {
            let rows: Vec<_> = memory_sites::report().into_iter().map(|r| serde_json::json!({
                "name":r.name,"phase":r.phase,"source_file":r.source_file,"source_line":r.source_line,
                "scope_kind":r.scope_kind,"requested_bytes":r.requested_bytes,
                "allocation_calls":r.allocation_calls,"observations":r.observations
            })).collect();
            fs::write(
                format!("{}-sites.json", prefix.display()),
                serde_json::to_vec_pretty(&rows)?,
            )?;
        }
        checkpoint("post_roots_retirement", memory())?;
        Ok(())
    })?;
    drop(inputs);
    checkpoint("post_retirement", memory())?;
    println!("{}", serde_json::json!({"complete": true}));
    Ok(())
}
